### ISSUE TITLE
Discrete slider 

### ISSUE THREAD

**Issue by Karel-van-de-Plassche** [bokeh/bokeh#6861]
I'm looking to create a discrete slider. Specifically, I need the following features:
- Manual setting of slider values (original just does min/max and auto-generates values in between)
- Display of pips (unavailable)
- Snapping to range values (unavailable)

[For example](dataslicer.qualikiz.com).

I prefer a 'native' solution, as linkage with external libraries (e.g. IonRangeSlider) breaks quite often.

[proof-of-concept](https://github.com/Karel-van-de-Plassche/bokeh/tree/discreteSlider)
[diff](https://github.com/bokeh/bokeh/compare/master...Karel-van-de-Plassche:discreteSlider)

Proof of concept in action:
``` Python
offset = DiscreteSlider(title="offset", value=0.0, range={'min':-0.1, '50%': 3.3333, 'max':5})
```
![image](https://user-images.githubusercontent.com/6404167/29978461-a819c03c-8f41-11e7-80fd-99bfcacd85d7.png)

Issues:
- pips overlap with other divs (can be solved manually by specifying `height`)
- pip text overlaps with large pips (might be an issue with noUiSlider itself)

To discuss:
- Setting pip settings is awkward (`pip_mode`, `pip_density`), maybe setting with a dict instead
- Setting of `range` is awkward, but I copied the noUiSlider interface 1-to-1
- Might not need a separate `DiscreteSlider` class, as most settings can be incorporated in `AbstractSlider`

**Comment by Karel-van-de-Plassche** [comment]
@mattpap and @bryevdv might want to chip in

**Comment by bryevdv** [comment]
@Karel-van-de-Plassche would you be interested in submitting a few PRs towards this? I think it would be good to break things up, i.e.

* a PR to add support for a "pips" dict to existing slider (just pass as-is to nouislider)
* another PR to add range snapping
* last PR to add support for manual values (need some discussion of API, new model vs adding to existing)

**Comment by Karel-van-de-Plassche** [comment]
I'd be happy to submit some PRs. I've opened separate issues as requested by BEP1. Let's discuss further there and close this for now.

**Comment by bearums** [comment]
Hello @Karel-van-de-Plassche, nice work! I would like to use a discrete slider in Bokeh but I am not able to implement your code into bokeh. Please could you give me a tip on how to do this? 

Ben

**Comment by Karel-van-de-Plassche** [comment]
Unfortunately this code is very outdated at this point. I'd advise to try extending Bokeh with the ion.rangeSlider, as in the examples: https://docs.bokeh.org/en/latest/docs/user_guide/extensions_gallery/widget.html#userguide-extensions-examples-widget, or jumping in the discussion here #7852.

**Comment by bearums** [comment]
Thanks for your response. Unfortunately I do not possess the skills to write my own custom widget. Instead I have made a poor man's version using FuncTickFormatter - note that this function is only available in versions > 2.0.2. 

For others who faced the same problem, here is my code 

```
from bokeh.models.formatters import  FuncTickFormatter

label_dict = {0:"apples", 1: "bananas", 2:"oranges"}

fmt= FuncTickFormatter(code="""
    var labels = %s;
    return labels[tick];
""" % label_dict)

slider = Slider(start=min(label_dict.keys()) ,end=max(label_dict.keys()), step=1, value=1, 
           format=fmt)
```
